### PR TITLE
Implement Strict global configuration support to control whether non-empty author attribute is allowed or not

### LINKED ISSUE #3873: Liquibase 4.19.0 now requires a non-empty author attribute
### Search first

- [X] I searched and no similar issues were found

### Description

Feature #1290 released in 4.19.0 newly enforces an author to be set. As we are currently passing empty strings (e.g. &lt;changeSet id="1" author=""&gt;), validation fails - even for change sets already applied to production. This is a backward incompatible change that should have been avoided.
At the moment I can see the only possible way to upgrade to the new version: to add authors everywhere, which violates time-proven rule to never touch applied migrations. And to clear checksums in existing databases which goes against equally proven principle: do not mess about with production data.
Is there a chance the feature will be reverted?

### Steps To Reproduce

Just update the library.

### Actual Behavior

_No response_

### Expected/Desired Behavior

Do not enforce authors. Or make the validation configurable.

### Liquibase Version

4.19.0

### Database Vendor & Version

_No response_

### Liquibase Integration

Spring Boot

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

_No response_

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### LINKED ISSUE #4053: #1290 introduced a breaking change that cannot be fixed/
### Search first

- [X] I searched and no similar issues were found

### Description

https://github.com/liquibase/liquibase/pull/3397 has introduced a breaking change. There is no way to fix the break or do a workaround. Liquibase 4.20.0 contains a security update that we are unable to update to.
Having the author enforced means that we could not update to 4.19.0 and 4.20.0. See our existing migrations.xml:
https://github.com/alphagov/pay-connector/blob/master/src/main/resources/migrations.xml
We have never included the author, left it blank as it was not a requirement. Enforcing the author breaks the md5sum of the change set and thus Liquibase thinks it's a new one.
Exception:
`liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset`

### Steps To Reproduce

Run a [migrations.xml](https://github.com/alphagov/pay-connector/blob/master/src/main/resources/migrations.xml) with Liquibase version 4.18.0. Then update to version 4.20.0. Try to run the migrations set again. It will break, saying that:
`ChangeSet Author is empty, migrations.xml::build gateway table::`. Add an author to the ChangeSets, for example author="joe.bog". Run migrations again, it will try to run all ChangeSets as the md5sum does not match.

### Expected/Desired Behavior

Expected behaviour is either do not enforce author for already ran changesets, or have the option to ignore empty author field.

### Liquibase Version

4.19.0

### Database Vendor & Version

PostGres 11.5

### Liquibase Integration

dropwizard-migrations 2.1.4

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

Ubuntu, MacOS

### Additional Context

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by MalloD12**
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Implement Strict global configuration support to control whether non-empty author attribute is allowed or not. In other words, when setting up `strict = true` empty author attribute won't be allowed and a validation error will be displayed, instead of deploy changeSet successfully when `strict = false`

Fixes #3873 
Fixes #4053

## Things to be aware of
- None

## Things to worry about
- None

## Additional Context
- Nothing

### CODE REVIEW

**Review by filipelautert — APPROVED** [review]

File: liquibase-core/src/main/java/liquibase/changelog/visitor/ValidatingVisitor.java [hunk 1]
```diff
@@ -168,7 +169,7 @@ private boolean areChangeSetAttributesValid(ChangeSet changeSet) {
         boolean valid = false;
         if (authorEmpty && idEmpty) {
             validationErrors.addError("ChangeSet Id and Author are empty", changeSet);
-        } else if (authorEmpty) {
+        } else if (authorEmpty && GlobalConfiguration.STRICT.getCurrentValue()) {
```
Comment (suryaaki2) [inline liquibase-core/src/main/java/liquibase/changelog/visitor/ValidatingVisitor.java:172]: Should the variable name be GlobalConfiguration.AUTHOR_STRICT?

**Review by suryaaki2 — CHANGES_REQUESTED** [review]

File: liquibase-core/src/main/java/liquibase/changelog/visitor/ValidatingVisitor.java (hunk 1 above)
Comment (filipelautert) [reply liquibase-core/src/main/java/liquibase/changelog/visitor/ValidatingVisitor.java:172]: Hello @suryaaki2 - it was discussed with Nathan and agreed to use the generic Strict flag.

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        changeSet.addChange(change)
+
+        ValidatingVisitor handler = new ValidatingVisitor(new ArrayList<RanChangeSet>())
+        handler.visit(changeSet, new DatabaseChangeLog(), new MockDatabase(), null)
+
+        then:
+        !handler.validationPassed()
+    }
+
+    void "validate successful visit with none dbms set"() throws Exception {
```
Comment (XDelphiGrl) [inline liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:237]: The literal strings "none" and "all" are valid DBMS values? The purpose of "all" seems relatively straightforward, if not a bit redundant as the default is "all" when no dbms attribute is configured. However, what does it mean to have a changeset that has `dbms="none"`?

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            void run() throws Exception {
+                handler.visit(changeSet, new DatabaseChangeLog(), null, null)
+            }
+        })
+
+        then:
+         handler.getValidationErrors().getErrorMessages().get(0).contains("ChangeSet Author is empty")
+    }
+
+    void "validate error on empty author when strict configuration is set as false"() {
```
Comment (XDelphiGrl) [inline liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:119]: A more accurate test description would be "validate *there is no* error on empty author when strict configuration is set as false".

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy
Comment (XDelphiGrl) [inline liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:1]: This is a very nice set of tests, @MalloD12!

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy (hunk 2 above)
Comment (MalloD12) [reply liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:237]: Hey Erz, yeap they are. At least this is what the documentation states for this:

> Specifies which database type(s) a changeset is to be used for. See valid database type names on [Supported Databases](https://www.liquibase.org/get-started/databases). Separate multiple databases with commas. Specify that a changeset is not applicable to a particular database type by prefixing with !. The keywords all and none are also available.

Also, these are existent tests that I just migrated from JUnit to Groovy after adding some new tests related to this PR.

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy (hunk 3 above)
Comment (MalloD12) [reply liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:119]: Sounds good to me, I'll make this update in a bit.

File: liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy
Comment (MalloD12) [reply liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:1]: No flowers to me regarding these tests. Here, I just added the ones related to empty author. The rest were existent unit tests that already exist in a JUnit format.

Comment (XDelphiGrl) [reply liquibase-core/src/test/groovy/liquibase/changelog/visitor/ValidatingVisitorTest.groovy:1]: All the flowers go to you, my friend!! Updating from JUnit is awesome. And the new tests for empty author are great; I feel confident the feature will work as expected.

**Review by suryaaki2 — APPROVED** [review]

**Comment by XDelphiGrl** [comment]
**Functional Test Failure Analysis**

* MSSQL Error: SQLCmd file extension issue. (KNOWN, no action necessary)
* MSSQL RDS: Unknown; report not saved on GH. (RERUN)
* Azure Postgres SS: Cannot get connection. (ENVIRONMENTAL, no action necessary)
* Postgres Error Rollback One changeset error on Maven 
```
utils.cli.CommandLineException: Unexpected error running Liquibase: 'liquibase.logging.mdc.customobjects.ChangesetsRolledback$ChangeSet liquibase.logging.mdc.customobjects.ChangesetsRolledback$ChangeSet.fromChangeSet(liquibase.changelog.ChangeSet)'
```

* MSSQL GCP: API mismatch
```
An API incompatibility was encountered while executing org.liquibase:liquibase-maven-plugin:0-SNAPSHOT:rollbackOneChangeSet: 
java.lang.NoSuchMethodError: 'java.lang.String com.datical.liquibase.ext.changelog.filter.SingleRanChangeSetFilter.getDisplayName()'
```

CC @MalloD12

**Comment by XDelphiGrl** [comment]
@MalloD12, I'm going to push this back over to Development. I can reproduce the Postgres failure locally. You can try for yourself with the test at line 55 (@Test(testName = "MVN/CLI. Rollback One ChangeSet", dataProvider = "changelog"). The description for the test is outdated; at some point, it was extended to all database platforms, but the description still mentions Oracle.

**Test Class**
src/test/java/tests/functional/rollback/rollbackOneChangeSet/RollbackOneChangeSetTests.java 

**Error Message**
```
Rolling Back Changeset:changelog.xml::createProcedure::Liquibase Pro User

Unexpected error running Liquibase: 'liquibase.logging.mdc.customobjects.ChangesetsRolledback$ChangeSet liquibase.logging.mdc.customobjects.ChangesetsRolledback$ChangeSet.fromChangeSet(liquibase.changelog.ChangeSet)'
```
**Failing Version**
OSS:  //fix/issue-3873/8738/1ad0af/
Commercial:  master/3237/426eca

**Comment by XDelphiGrl** [comment]
@MalloD12, I'm going to do some research to answer your question.

**Review by XDelphiGrl — APPROVED** [review]
Defining a changeset author is useful for traceability and auditing but was never required. Liquibase 4.20 broke backwards-compatibility by enforcing a requirement that changesets have a non-empty `author` attribute. This PR restores the legacy behavior for such changesets.
* Although the value for `author` can be empty, the attribute itself _must_ be specified; this has always been required (even prior to 4.20)
* If you prefer changesets with empty `author` to throw a validation error, execute validate with the `--strict true` global parameter. For example:
  * `liquibase --strict true validate --changelog-file myemptyauthorchangelog.xml`
* Test harness execution passing.
* There were two functional test failures; one for a known test issue and another due to a build issue. 
* New unit tests added for empty `author` AND  a bonus update from Java to Groovy for the test class.
* Manual test of `validate` and `update` commands with `strict=true` and `strict=false` passing.


APPROVED

---

**Example Changeset with Empty Author**
```
	<changeSet author=""  id="1::emptyAuthor" labels="emptyAuthor">
        <createTable tableName="emptyauthor">
            <column name="name" type="CHAR(20)"/>
        </createTable>
    </changeSet>
```

** Validate Command Output for Empty Author Changeset**

| Liquibase Version | strict setting | validation outcome |
| ------------------ | ------------ | --------------------|
| 4.19 and earlier | default (false) | no validation error |
| 4.20 | default (false) | validation error due to empty author |
| 4.21 | default (false) | no validation error |
| 4.21 | true | validation error due to empty author |
